package/zmq_adapter/src/zmq_adapter.c

```diff
@@ -786,104 +705,132 @@ static void io_loop_pubsub(handle_t *read_handle, handle_t *write_handle)
-      continue;
-    } else if (poll_ret < 0) {
-      /* Break on error */
-      break;
-    }
-
-    if (poll_ret == 0) {
-      /* Timeout */
-      if ((rep_handle->zsock != NULL) && reply_pending) {
-        /* Assume the outstanding request was lost.
-         * Reset the REP socket so that another request may be received. */
-        syslog(LOG_ERR, "reply timeout - resetting socket");
-        zsock_restart(&rep_handle->zsock);
-        if (rep_handle->zsock == NULL) {
-          break;
-        }
-        reply_pending = false;
+static void req_handle_callback(pk_loop_t *loop, void *handle, void *context)
+{
+  reqrep_state_t *ctx = (reqrep_state_t *)context;
+
+  if (!ctx->reply_pending) {
+    syslog(LOG_ERR, "warning: reply received but not pending");
+    if (ctx->rep_handle->pk_ept != NULL) {
+      /* Reply received with no request outstanding.
+       * Read and drop data from req_handle. */
+      syslog(LOG_ERR, "dropping data");
+      uint8_t buffer[READ_BUFFER_SIZE];
+      ssize_t read_count = handle_read(ctx->req_handle, buffer, sizeof(buffer));
+      debug_printf("read %zd bytes\n", read_count);
+      if (read_count <= 0) {
+        pk_loop_stop(loop);
       }
-      continue;
+      return;
     }
+  }
 
-    /* Check req_handle */
-    if (pollitems[POLLITEM_REQ].revents & ZMQ_POLLIN) {
-      if (!reply_pending) {
-        syslog(LOG_ERR, "warning: reply received but not pending");
-        if (rep_handle->zsock != NULL) {
-          /* Reply received with no request outstanding.
-           * Read and drop data from req_handle. */
-          syslog(LOG_ERR, "dropping data");
-          uint8_t buffer[READ_BUFFER_SIZE];
-          ssize_t read_count = handle_read(req_handle, buffer, sizeof(buffer));
-          debug_printf("read %zd bytes\n", read_count);
-          if (read_count <= 0) {
-            break;
-          }
+  bool ok;
+  if (frame_transfer(ctx->req_handle, ctx->rep_handle, &ok) <= 0) {
+    pk_loop_stop(loop);
+  }
 
-          continue;
-        }
-      }
+  if (ok) {
+    ctx->reply_pending = false;
+  }
+}
 
-      bool ok;
-      if (frame_transfer(req_handle, rep_handle, &ok) <= 0) {
-        break;
-      }
+static void timeout_callback(pk_loop_t *loop, void *handle, void *context)
+{
+  reqrep_state_t *ctx = (reqrep_state_t *)context;
+
+  /* Assume the outstanding request was lost.
+   * Reset the REP socket so that another request may be received. */
+  syslog(LOG_ERR, "reply timeout - resetting socket");
+  pk_endpoint_restart(&ctx->rep_handle->pk_ept);
+  if (ctx->rep_handle->pk_ept == NULL) {
+    pk_loop_stop(loop);
+  }
+  ctx->reply_pending = false;
+
+  // remove timer or this will continue to trigger
+  pk_loop_remove_handle(handle);
+  ctx->timeout_handle = NULL;
+}
 
-      if (ok) {
-        reply_pending = false;
+static void rep_handle_callback(pk_loop_t *loop, void *handle, void *context)
+{
+  (void)handle;
+  reqrep_state_t *ctx = (reqrep_state_t *)context;
```

Comment:
We should consider dropping req/rep support for now, it's not used anywhere.

Reply:
Agreed. I'll remove it.